### PR TITLE
chore: introduce `libp2p-test-utils`

### LINKED ISSUE #4992: `tracing_subscriber` should log to stderr
### Summary

Diagnostics (e.g. logging) should go to stderr. Output of a program should go to stdout.

### Expected behavior

`tracing_subscriber` to log to stderr.

### Actual behavior

`tracing_subscriber` is logging to stdout.

Upstream bug report https://github.com/tokio-rs/tracing/issues/2492

### Relevant log output

_No response_

### Possible Solution

One can easily override the default:

``` rust
    let _ = tracing_subscriber::fmt()
        .with_env_filter(EnvFilter::from_default_env())
        .with_writer(std::io::stderr)
        .try_init();
```

Initialization of `tracing_subscriber` is already quite noisy. E.g. it requires 4 lines instead of the 1 line of `env_logger`. Should we introduce some abstraction?

### Version

`tracing_subscriber` introduced with https://github.com/libp2p/rust-libp2p/pull/4282.

### Would you like to work on fixing this bug ?

No

### PR DESCRIPTION
**Pull request by kamuik16**
## Description

Fixes #4992 

Based on the conversation in #4992.

## Change checklist
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: misc/logging/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "libp2p-logging"
+version = "0.1.0"
+edition = "2021"
+authors = ["Krishang Shah <shahkrishang11@gmail.com>"]
+license = "MIT"
+repository = "https://github.com/libp2p/rust-libp2p"
+publish = false
```
Comment (jxs) [inline misc/logging/Cargo.toml]: if this is set to false we cannot use it on the rest of the code

File: misc/logging/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+use tracing_subscriber::EnvFilter;
+
+pub fn init_tracing_subscriber_with_default_env_filter() {
+    let _ = tracing_subscriber::fmt()
```
Comment (jxs) [inline misc/logging/src/lib.rs]: we can just call `init_tracing_subscriber_with_env_filter` with `EnvFilter::from_default_env()` right?

**Review by jxs — COMMENTED** [review]
Thanks for this.
Overall looks good to me, left some minor comments. We also need `CHANGELOG.md` entries.
 I think when Thomas mentioned the `publish = false` wasn't accounting what he said later on https://github.com/libp2p/rust-libp2p/issues/4992#issuecomment-1865377818

File: misc/logging/Cargo.toml (hunk 1 above)
Comment (kamuik16) [reply misc/logging/Cargo.toml]: > if this is set to false we cannot use it on the rest of the code

Right now, it is only used as a `dev-dependency` everywhere other than `protocols/perf` and `misc/server`.
If we only use it as a `dev-dependency`, we don't need to publish it. So I think it is best not to publish it and I can remove the usage of `libp2p-logging` from `protocols/perf` and `misc/server`. wdyt?

File: misc/logging/src/lib.rs (hunk 2 above)
Comment (kamuik16) [reply misc/logging/src/lib.rs]: > we can just call `init_tracing_subscriber_with_env_filter` with `EnvFilter::from_default_env()` right?

yes, but `init_tracing_subscriber_with_default_env_filter` seemes neater. Can change if neede, but this way it will only take up one fn call without having to give any input and importing the `EnvFilter` to do `EnvFilter::from_default_env()`.

File: misc/logging/Cargo.toml (hunk 1 above)
Comment (kamuik16) [reply misc/logging/Cargo.toml]: or am I missing something. I could be wrong though :)

**Comment by kamuik16** [comment]
> Thanks for this. Overall looks good to me, left some minor comments. We also need `CHANGELOG.md` entries. I think when Thomas mentioned the `publish = false` wasn't accounting what he said later on [#4992 (comment)](https://github.com/libp2p/rust-libp2p/issues/4992#issuecomment-1865377818)

> We also need `CHANGELOG.md` entries.

Which ones do you suggest? Since the logging is only used in testing part and is not the part of public API, I don't think it needs `CHANGELOG.md` entries.

File: misc/logging/src/lib.rs (hunk 2 above)
Comment (jxs) [reply misc/logging/src/lib.rs]: what I meant is that this function, `init_tracing_subscriber_with_default_env_filter` inside can just call `init_tracing_subscriber_with_env_filter` with `EnvFilter::from_default_env()`  instead of duplicating code

File: misc/logging/Cargo.toml (hunk 1 above)
Comment (jxs) [reply misc/logging/Cargo.toml]: yeah but both `protocols/perf` and `misc/server` are published, see [1](https://crates.io/crates/libp2p-perf) and [2](https://crates.io/crates/libp2p-server)

**Comment by jxs** [comment]
> > Thanks for this. Overall looks good to me, left some minor comments. We also need `CHANGELOG.md` entries. I think when Thomas mentioned the `publish = false` wasn't accounting what he said later on [#4992 (comment)](https://github.com/libp2p/rust-libp2p/issues/4992#issuecomment-1865377818)
> 
> > We also need `CHANGELOG.md` entries.
> 
> Which ones do you suggest? Since the logging is only used in testing part and is not the part of public API, I don't think it needs `CHANGELOG.md` entries.

the ones we publish, `libp2p-perf` and `libp2p-server`

File: misc/logging/Cargo.toml (hunk 1 above)
Comment (kamuik16) [reply misc/logging/Cargo.toml]: > yeah but both `protocols/perf` and `misc/server` are published, see [1](https://crates.io/crates/libp2p-perf) and [2](https://crates.io/crates/libp2p-server)

done

File: misc/logging/src/lib.rs (hunk 2 above)
Comment (kamuik16) [reply misc/logging/src/lib.rs]: > what I meant is that this function, `init_tracing_subscriber_with_default_env_filter` inside can just call `init_tracing_subscriber_with_env_filter` with `EnvFilter::from_default_env()` instead of duplicating code

ahh, yes. done.

**Comment by kamuik16** [comment]
> > > Thanks for this. Overall looks good to me, left some minor comments. We also need `CHANGELOG.md` entries. I think when Thomas mentioned the `publish = false` wasn't accounting what he said later on [#4992 (comment)](https://github.com/libp2p/rust-libp2p/issues/4992#issuecomment-1865377818)
> > 
> > 
> > > We also need `CHANGELOG.md` entries.
> > 
> > 
> > Which ones do you suggest? Since the logging is only used in testing part and is not the part of public API, I don't think it needs `CHANGELOG.md` entries.
> 
> the ones we publish, `libp2p-perf` and `libp2p-server`

added `CHANGELOG.md` entries.

**Comment by kamuik16** [comment]
@jxs would appreciate your further review here :)

**Comment by kamuik16** [comment]
`semver` check is just failing because the libp2p-logging crate has not been published yet.

**Review by jxs — APPROVED** [review]
LGTM, thanks Krishang, thanks for the ping! 
Giving more thought about this one taking into account we'll have to release it and be yet another crate to maintain in exchange for duplicating a couple lines across the repo it's not clear its value, but I am willing to move forward, wdyt @elenaf9 @dariusc93 @guillaumemichel @drHuangMHT

**Comment by kamuik16** [comment]
> LGTM, thanks Krishang, thanks for the ping! Giving more thought about this one taking into account we'll have to release it and be yet another crate to maintain in exchange for duplicating a couple lines across the repo it's not clear its value, but I am willing to move forward, wdyt @elenaf9 @dariusc93 @guillaumemichel @drHuangMHT

yeah, that's why I was suggesting to keep the `publish = false` and remove its usage from the packages that are released.

Majority of the usage of this is in the tests, so no need to publish it.

**Comment by jxs** [comment]
> > LGTM, thanks Krishang, thanks for the ping! Giving more thought about this one taking into account we'll have to release it and be yet another crate to maintain in exchange for duplicating a couple lines across the repo it's not clear its value, but I am willing to move forward, wdyt @elenaf9 @dariusc93 @guillaumemichel @drHuangMHT
> 
> yeah, that's why I was suggesting to keep the `publish = false` and remove its usage from the packages that are released.
> 
> Majority of the usage of this is in the tests, so no need to publish it.

but then we still need to have duplicated log initialization code on the published crates (`libp2p-perf` and `libp2p-server`) right? Or how would you address it? Nonetheless thanks for your patience bearing with the process Krishang.

File: misc/logging/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+use tracing_subscriber::EnvFilter;
+
+pub fn init_tracing_subscriber_with_default_env_filter() {
+    init_tracing_subscriber_with_env_filter(EnvFilter::from_default_env());
+}
+
+pub fn init_tracing_subscriber_with_env_filter(filter: impl Into<EnvFilter>) {
```
Comment (elenaf9) [inline misc/logging/src/lib.rs]: Nit: can we shorten these names, e.g. just `with_{default_}env_filter`?

I think the test etc where we use this don't really need to know about the inner logging system. Especially because there are no alternatives to `tracing_subscriber` implemented in this crate.

File: misc/logging/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+use tracing_subscriber::EnvFilter;
```
Comment (elenaf9) [inline misc/logging/src/lib.rs]:
```suggestion
pub use tracing_subscriber::EnvFilter;
```

Re-export `EnvFilter` so that no additional dependency to `tracing_subscriber` is needed when using `init_tracing_subscriber_with_env_filter`?

File: misc/logging/src/lib.rs (hunk 3 above)
Comment (elenaf9) [inline misc/logging/src/lib.rs]:
```suggestion
pub fn init_tracing_subscriber_with_env_filter(filter: EnvFilter) {
```

Nit: Is there any type that implements `Into<EnvFilter>` other than `EnvFilter` itself?

**Review by elenaf9 — COMMENTED** [review]
> yeah, that's why I was suggesting to keep the publish = false and remove its usage from the packages that are released.
>
> Majority of the usage of this is in the tests, so no need to publish it.

I am also leaning towards not publishing `libp2p-logging`, and instead duplicating the lines in the two published crates.
It's not ideal, but it still:
- avoids the verbose `let _ = tracing_subscriber::fmt()...` in all other tests/ examples
- doesn't require us to maintain yet another published crate.

File: misc/logging/src/lib.rs (hunk 3 above)
Comment (kamuik16) [reply misc/logging/src/lib.rs]: > Nit: Is there any type that implements `Into<EnvFilter>` other than `EnvFilter` itself?

yes, string and string literals like "my_crate=info,my_crate::my_mod=debug,[my_span]=trace".

Comment (kamuik16) [reply misc/logging/src/lib.rs]: okay

File: misc/logging/src/lib.rs (hunk 4 above)
Comment (kamuik16) [reply misc/logging/src/lib.rs]: done

File: misc/logging/src/lib.rs (hunk 3 above)
Comment (kamuik16) [reply misc/logging/src/lib.rs]: done

**Comment by kamuik16** [comment]
> > yeah, that's why I was suggesting to keep the publish = false and remove its usage from the packages that are released.
> > Majority of the usage of this is in the tests, so no need to publish it.
> 
> I am also leaning towards not publishing `libp2p-logging`, and instead duplicating the lines in the two published crates. It's not ideal, but it still:
> 
> * avoids the verbose `let _ = tracing_subscriber::fmt()...` in all other tests/ examples
> * doesn't require us to maintain yet another published crate.

yep, I do agree on this, it's better to have 2-3 lines of duplicated code rather than maintaining a whole crate.

I have reverted back the changes on the published crates and made `libp2p-logging` with `publish = false`.

File: misc/logging/CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+## 0.1.0
+
+- Introduce 'libp2p-logging` crate.
+  See [PR 5725](https://github.com/libp2p/rust-libp2p/pull/5725).
```
Comment (elenaf9) [inline misc/logging/CHANGELOG.md]: Do we need a changelog if the crate isn't published?

File: misc/multistream-select/src/dialer_select.rs [hunk 6]
```diff
@@ -275,13 +275,11 @@ mod tests {
...
-                    EnvFilter::builder()
-                        .with_default_directive(LevelFilter::DEBUG.into())
-                        .from_env_lossy(),
-                )
-                .try_init();
+            libp2p_logging::with_env_filter(
+                EnvFilter::builder()
+                    .with_default_directive(LevelFilter::DEBUG.into())
+                    .from_env_lossy(),
+            );
```
Comment (elenaf9) [inline misc/multistream-select/src/dialer_select.rs]: Use `libp2p_logging::EnvFilter` and remove the dependency to `tracing_subscriber`.

File: protocols/stream/tests/lib.rs [hunk 7]
```diff
@@ -13,15 +13,12 @@ const PROTOCOL: StreamProtocol = StreamProtocol::new("/test");
...
-                .unwrap(),
-        )
-        .with_test_writer()
-        .try_init();
+    libp2p_logging::with_env_filter(
+        EnvFilter::builder()
+            .with_default_directive(LevelFilter::DEBUG.into())
+            .from_env()
+            .unwrap(),
+    );
```
Comment (elenaf9) [inline protocols/stream/tests/lib.rs]: Same here. Use `libp2p_logging::EnvFilter` and remove the dependency to `tracing_subscriber`.

File: swarm/src/lib.rs [hunk 8]
```diff
@@ -1497,14 +1497,17 @@ impl Config {
...
     ///
     /// A connection is considered idle if:
     /// - There are no active inbound streams.
     /// - There are no active outbounds streams.
-    /// - There are no pending outbound streams (i.e. all streams requested via [`ConnectionHandlerEvent::OutboundSubstreamRequest`] have completed).
-    /// - Every [`ConnectionHandler`] returns `false` from [`ConnectionHandler::connection_keep_alive`].
+    /// - There are no pending outbound streams (i.e. all streams requested via
+    ///   [`ConnectionHandlerEvent::OutboundSubstreamRequest`] have completed).
+    /// - Every [`ConnectionHandler`] returns `false` from
+    ///   [`ConnectionHandler::connection_keep_alive`].
```
Comment (elenaf9) [inline swarm/src/lib.rs]: Please revert unrelated formatting changes.

File: misc/logging/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+pub fn with_default_env_filter() {
+    with_env_filter(EnvFilter::from_default_env());
+}
+
+pub fn with_env_filter(filter: impl Into<EnvFilter>) {
+    let _ = tracing_subscriber::fmt()
+        .with_env_filter(filter)
+        .with_writer(std::io::stderr)
+        .try_init();
+}
```
Comment (elenaf9) [inline misc/logging/src/lib.rs]: Please add a short description to each function :pray:

**Review by elenaf9 — COMMENTED** [review]
Thanks for the follow-up @kamuik16!

I just noticed that the new `libp2p_logging` crate is not used by any crates in `examples/`. Was that on purpose / is there a reason why it can't be used there?

**Comment by kamuik16** [comment]
> Thanks for the follow-up @kamuik16!
> 
> I just noticed that the new `libp2p_logging` crate is not used by any crates in `examples/`. Was that on purpose / is there a reason why it can't be used there?

yeah, that was on purpose, examples are meant for user to copy paste and run. If we don't publish the crate, examples would compile in our repository but users who are still new and just want to copy paste and run them, won't be able to do so.

does this make sense?

File: misc/logging/CHANGELOG.md (hunk 5 above)
Comment (kamuik16) [reply misc/logging/CHANGELOG.md]: I think it's not very bad to have the changelog even if we don't publish it. We can look at its history with it. Can be removed if wanted.

File: swarm/src/lib.rs (hunk 8 above)
Comment (kamuik16) [reply swarm/src/lib.rs]: > Please revert unrelated formatting changes.

these are just formatting changes I got on running `cargo +nightly fmt` which uses `rustfmt.toml` features. Someone must not have used it on a PR that was merged previously. When I merged master into my branch and ran, it automatically did this.

File: misc/logging/src/lib.rs (hunk 9 above)
Comment (kamuik16) [reply misc/logging/src/lib.rs]: yeah

File: protocols/stream/tests/lib.rs (hunk 7 above)
Comment (kamuik16) [reply protocols/stream/tests/lib.rs]: done

File: misc/multistream-select/src/dialer_select.rs (hunk 6 above)
Comment (kamuik16) [reply misc/multistream-select/src/dialer_select.rs]: done

File: swarm/src/lib.rs (hunk 8 above)
Comment (elenaf9) [reply swarm/src/lib.rs]: Extracted into #5742.

**Comment by elenaf9** [comment]
> yeah, that was on purpose, examples are meant for user to copy paste and run. If we don't publish the crate, examples would compile in our repository but users who are still new and just want to copy paste and run them, won't be able to do so.
> 
> does this make sense?

Yes, makes sense, although it's a bit unfortunate. I didn't think about that before and based on the discussion in #4992 assumed we would be using it for both tests and examples. My bad.

I would still go ahead with the current PR. Wdyt @jxs?

Another thought I just had is that if this is only used in tests, we could consider generalizing `libp2p-logging` into `test-utils` / `test-common` or sth like that, so next time we have another function or logic that is commonly reused in tests we can just add it there.

**Comment by kamuik16** [comment]
> > yeah, that was on purpose, examples are meant for user to copy paste and run. If we don't publish the crate, examples would compile in our repository but users who are still new and just want to copy paste and run them, won't be able to do so.
> > does this make sense?
> 
> Yes, makes sense, although it's a bit unfortunate. I didn't think about that before and based on the discussion in #4992 assumed we would be using it for both tests and examples. My bad.
> 
> I would still go ahead with the current PR. Wdyt @jxs?
> 
> Another thought I just had is that if this is only used in tests, we could consider generalizing `libp2p-logging` into `test-utils` / `test-common` or sth like that, so next time we have another function or logic that is commonly reused in tests we can just add it there.

seems reasonable

**Comment by jxs** [comment]
> > yeah, that was on purpose, examples are meant for user to copy paste and run. If we don't publish the crate, examples would compile in our repository but users who are still new and just want to copy paste and run them, won't be able to do so.
> > does this make sense?
> 
> Yes, makes sense, although it's a bit unfortunate. I didn't think about that before and based on the discussion in #4992 assumed we would be using it for both tests and examples. My bad.
> 
> I would still go ahead with the current PR. Wdyt @jxs?
> 
> Another thought I just had is that if this is only used in tests, we could consider generalizing `libp2p-logging` into `test-utils` / `test-common` or sth like that, so next time we have another function or logic that is commonly reused in tests we can just add it there.

yeah makes sense to me, can we then update this PR Krishang and do it? Again, thanks for the patience! :heart:

**Comment by kamuik16** [comment]
> > > yeah, that was on purpose, examples are meant for user to copy paste and run. If we don't publish the crate, examples would compile in our repository but users who are still new and just want to copy paste and run them, won't be able to do so.
> > > does this make sense?
> > 
> > 
> > Yes, makes sense, although it's a bit unfortunate. I didn't think about that before and based on the discussion in #4992 assumed we would be using it for both tests and examples. My bad.
> > I would still go ahead with the current PR. Wdyt @jxs?
> > Another thought I just had is that if this is only used in tests, we could consider generalizing `libp2p-logging` into `test-utils` / `test-common` or sth like that, so next time we have another function or logic that is commonly reused in tests we can just add it there.
> 
> yeah makes sense to me, can we then update this PR Krishang and do it? Again, thanks for the patience! ❤️

Hey @jxs, no worries. I updated the PR to move the `libp2p-logging` to `libp2p-test-utils`. Also changing the PR title now.

**Review by elenaf9 — APPROVED** [review]
Thank you for your patience @kamuik16! LGTM

**Review by jxs — APPROVED** [review]
Thanks Krishang! LGTM Can you do one last chore and address the merge conflicts? :grin:

**Review by dariusc93 — APPROVED** [review]

**Comment by kamuik16** [comment]
> Thanks Krishang! LGTM Can you do one last chore and address the merge conflicts? 😁

resolved @jxs 😄

**Review by jxs — APPROVED** [review]

**Comment by jxs** [comment]
Thanks!
